src/engine/Sort.cpp

```diff
@@ -52,30 +55,131 @@ std::string Sort::getDescriptor() const {
+  // directory.
+  const std::string& onDiskBase =
+      getExecutionContext()->getIndex().getOnDiskBase();
+  std::ostringstream oss;
+  oss << onDiskBase << ".sort-" << reinterpret_cast<uintptr_t>(this) << "-"
+      << std::this_thread::get_id() << ".dat";
+  std::string tempFilename = oss.str();
+
+  // Get memory limit from the same runtime parameters that we use for the
+  // materialized view writer.
+  //
+  // TODO: Give this runime parameter a more general name.
+  ad_utility::MemorySize memoryLimit =
+      getRuntimeParameter<&RuntimeParameters::materializedViewWriterMemory_>();
+
+  // Get the number of columns from the subtree (this works for both lazy and
+  // materialized results).
+  size_t numColumns = subtree_->getResultWidth();
+
+  // Create the external sorter with our dynamic column comparator. We use a
+  // shared pointer so the sorter survives when the lambda below captures it.
+  using Sorter = ad_utility::CompressedExternalIdTableSorter<SortByColumns, 0>;
+  auto sorter = std::make_shared<Sorter>(
+      tempFilename, numColumns, memoryLimit, allocator(),
+      ad_utility::DEFAULT_BLOCKSIZE_EXTERNAL_ID_TABLE,
+      SortByColumns{sortColumnIndices_});
+
+  // Push all rows into the sorter, handling both lazy and materialized results.
+  // For materialized results, we copy the local vocab, for lazy results we
+  // merge the local vocabs from the individual blocks.
+  auto mergedLocalVocab = std::make_shared<LocalVocab>();
+  if (subRes->isFullyMaterialized()) {
+    sorter->pushBlock(subRes->idTable());
+    *mergedLocalVocab = subRes->getCopyOfLocalVocab();
+  } else {
+    for (Result::IdTableVocabPair& pair : subRes->idTables()) {
+      checkCancellation();
+      sorter->pushBlock(pair.idTable_);
+      mergedLocalVocab->mergeWith(pair.localVocab_);
+    }
+  }
+
+  // If laziness is not requested, materialize the result.
+  if (!requestLaziness) {
+    IdTable result{numColumns, allocator()};
+    for (auto& block : sorter->getSortedBlocks<0>()) {
+      checkCancellation();
+      result.insertAtEnd(block);
+    }
+    cancellationHandle_->resetWatchDogState();
+    checkCancellation();
+    return {std::move(result), resultSortedOn(), std::move(*mergedLocalVocab)};
+  }
+
+  // Otherwise, return a lazy result that yields sorted blocks. The first block
+  // gets the merged local vocab, subsequent blocks get empty local vocabs.
+  auto sortedBlocks = sorter->getSortedBlocks<0>();
+  bool isFirst = true;
+  return {Result::LazyResult{
+              ad_utility::OwningView{ad_utility::CachingTransformInputRange{
+                  std::move(sortedBlocks),
+                  [sorter, mergedLocalVocab, isFirst,
+                   this](IdTableStatic<0>& block) mutable {
+                    checkCancellation();
+                    IdTable table = std::move(block).toDynamic();
+                    LocalVocab localVocab =
+                        isFirst ? std::move(*mergedLocalVocab) : LocalVocab{};
```

Comment:
I don't think this is right. We have to always copy the local vocab for every `IdTableVocabPair` if the associated `IdTable` can contain a local vocab entry.

Reply:
You mean because a subsequent operation may just consume a subset of the blocks?

I guess you are right. What do you suggest?

Reply:
That's what I did now